### ISSUE TITLE
Data Science Preprocessing

### ISSUE THREAD

**Issue by kurtzace** [kurtzace/Diary2021#5]
## Eye ball the data using excel/Libre Calc, scroll up/down
Check
- Pattern of data like how dates are written, 
- Check how data entry people are recording NA values - 'NO CLUE', 'N/A', '0', 'UNKNOWN',......
- comments like refer to urls
- TODO comments
- Odd things like numeric names
- Country codes or full country names
- PII data like name/age/ssn/dob/blood group

## Load data to pandas
[pima diabetes data](https://www.kaggle.com/uciml/pima-indians-diabetes-database?select=diabetes.csv)
```
import pandas as pd
# Read
df = pd.read_csv("../input/pima-indians-diabetes-database/diabetes.csv")
# Looking at the first 5 rows of the data set
df.head()
```
![image](https://user-images.githubusercontent.com/2136211/110196151-ab442300-7e68-11eb-97ae-016136eb417f.png)

where
-  BMI: Body mass index 
- Outcome:  diabetes or not

I think if you eyeballed the NA values well. Load the data with eyeballed NA values.
```
na_values = ['NO CLUE', 'N/A', '0', 'UNKNOWN']
df = pd.read_csv("../input/pima-indians-diabetes-database/diabetes.csv", na_values=na_values)
```

### Nulls and datatypes
```
df.info()
```
It will give info like Non-Null Count values, data types, total rows count.

### total nulls
```
df.isnull().sum()

```

### NAN based on domain knowledge
Things like "Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI" cannot be 0 

treat them like nulls
```
df[["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]] = \
    df[["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]].replace(0, np.NaN)
```

### get a feel of count | mean | std | min | 25% | 50% | 75% | max
```
# Looking at the descriptive statistics of the data set
df.describe().T
```
## Rescale Data / Standardization / Feature scaling
- All attributes to have the same scale - important for machine learning - especially gradient descent.
- We dont want ML to treat 'money' (ranges 100s to 1000s) to be more important than 'v-engines' ranges 1 to 6
- allows us to reuse 'car' model for a 'cycle' too
```
from sklearn.preprocessing import MinMaxScaler 
scaler = MinMaxScaler(feature_range=(0, 1)) 
rescaledX = scaler.fit_transform(Input) 
```

### Standarization
or [standardization](https://user-images.githubusercontent.com/2136211/110197119-5b1c8f00-7e6f-11eb-9339-dcf0db5578a7.png)
![image](https://user-images.githubusercontent.com/2136211/110197143-8b642d80-7e6f-11eb-884d-1cde7e4bb6c5.png)
transform attributes with a Gaussian distribution and differing means and standard deviations to a standard Gaussian distribution with a mean of 0 and a standard deviation of 1.

```
# Python code to Standardize data (0 mean, 1 stdev) 
from sklearn.preprocessing import StandardScaler 
scaler = StandardScaler().fit(X) 
rescaledX = scaler.transform(X) 
```

### PCA
combine various variables into one
![image](https://user-images.githubusercontent.com/2136211/110197161-b189cd80-7e6f-11eb-8e93-cf6af84c06c8.png)

## Whitening
Remove un -correlated data 

Example - pressure in your pressure cooker is not related to housing prices.

### Binary encoding
```
from sklearn.preprocessing import Binarizer 
binarizer = Binarizer(threshold=0.0).fit(X) 
binaryX = binarizer.transform(input) 
```
## Appendix
1. https://hackernoon.com/what-steps-should-one-take-while-doing-data-preprocessing-502c993e1caa
2. https://www.udemy.com/course/the-data-science-course-complete-data-science-bootcamp/
3. https://towardsdatascience.com/data-preprocessing-concepts-fa946d11c825
4. https://www.upgrad.com/blog/data-preprocessing-in-machine-learning/
5. https://www.geeksforgeeks.org/data-preprocessing-machine-learning-python/
6. https://www.analyticsvidhya.com/blog/2016/07/practical-guide-data-preprocessing-python-scikit-learn/
7. https://www.xenonstack.com/blog/data-preprocessing-data-wrangling-in-machine-learning-deep-learning

## starter datasets
[pima diabetes data](https://www.kaggle.com/uciml/pima-indians-diabetes-database?select=diabetes.csv)

**Comment by kurtzace** [comment]
# Image pre processing

## Image mining
[reference](https://app.pluralsight.com/library/courses/mining-data-images/table-of-contents)

### Basic image reading /writing
```
import cv2 
img = cv2.imread(image_path) 
cv2.imwrite(filename, img) 
```

### Apply Perspective Transform Algorithm 
    matrix = cv2.getPerspectiveTransform(pts1, pts2) 
### Wrap the transformed image 
    result = cv2.warpPerspective(frame, matrix, (500, 600)) 

### Access pixel value
![image](https://user-images.githubusercontent.com/2136211/110204428-f24a0c80-7e98-11eb-8ff7-9519cfd504fe.png)
![image](https://user-images.githubusercontent.com/2136211/110204469-26253200-7e99-11eb-9d96-b3bc3e7470e2.png)

### cv2.selectROI can be used to select region of interest using mouse

### covert image to black and white
frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

### compare images using histogram
for forged cheques/signatures/counterfeit money
```
hist = cv2.calcHist([frame],[0],None,[4096],[0,4096])
cv2.normalize(hist, hist, alpha=0, beta=1, norm_type=cv2.NORM_MINMAX)
correl = cv2.compareHist(files_hist[f1], files_hist[f2], cv2.HISTCMP_CORREL)
if correl >= 0.999:
        print('MATCH:', correl, f1, f2)
```
for strict comparison
cv2.substract

### feature detection
![image](https://user-images.githubusercontent.com/2136211/110204765-de9fa580-7e9a-11eb-9112-40b318b0f500.png)


## Others
- Resize all images to consistent size 
[as seen here ](https://www.kaggle.com/iafoss/image-preprocessing-128x128) uses intelligent way to preserve aspect ratio and then perform 
```
 cv2.resize(img,(size,size)) 

```

- For medical images
[this one makes internal / external markers](https://www.kaggle.com/digvijayyadav/lung-fibrosis-dicom-image-preprocessing#Marker-controlled-Watershed-Transformation)
- [use kmeans to separate bone mass from lung/air](https://www.kaggle.com/digvijayyadav/lung-fibrosis-dicom-image-preprocessing)
- [read image metadata using pydicom.read_file(path)](https://www.kaggle.com/digvijayyadav/lung-fibrosis-dicom-image-preprocessing#Extract-Metadata)
- [Automatically identify the boundaries of the car in an image using opencv](https://www.kaggle.com/c/carvana-image-masking-challenge/discussion/39393)

**Comment by kurtzace** [comment]
## Vif

https://www.statsmodels.org/stable/generated/statsmodels.stats.outliers_influence.variance_inflation_factor.html

https://statisticalhorizons.com/multicollinearity

snippet from above
```
obtaining the R2 from that regression. The VIF is just 1/(1-(R^2)).
Estimates how much the variance of a coefficient is “inflated” because of linear dependence with other predictors. Thus, a VIF of 1.8 tells us that the variance (the square of the standard error) of a particular coefficient is 80% larger than it would be if that predictor was completely uncorrelated with all the other predictors.

The VIF has a lower bound of 1 but no upper bound. Authorities differ on how high the VIF has to be to constitute a problem. Personally, I tend to get concerned when a VIF is greater than 2.50, which corresponds to an R2 of .60 with the other variables.

above talks about three situations in which a high VIF is not a problem and can be safely ignored:
-  (dummy) variables
-  If you specify a regression model with both x and x2, there’s a good chance that those two variables will be highly correlated. 
-  high VIFs are control variables: the sample consists of U.S. colleges, the dependent variable is graduation rate, and the variable of interest is an indicator (dummy) for public vs. private. Two control variables are average SAT scores and average ACT scores for entering freshmen. 
```
in general

![image](https://user-images.githubusercontent.com/2136211/118392467-871a7480-b657-11eb-95de-261274ff2865.png)

**Comment by kurtzace** [comment]
## when data-frame is having no label
data = data.drop(['Unnamed: 0'], axis = 1)

**Comment by kurtzace** [comment]
## Pandas tips
Creating Checkpoints while Coding in Jupyter
```
df_checkpoint = df_concatenated.copy()
or
df_preprocessed.to_csv('Absenteeism_preprocessed.csv', index=False).

```
Concatenate the df_no_age and age_dummies

df_concatenated = pd.concat([df_no_age, age_dummies], axis=1)

**Comment by kurtzace** [comment]
### sort min,max
```
# df['Reason for Absence'].min()
# df['Reason for Absence'].max()
# pd.unique(df['Reason for Absence'])
# df['Reason for Absence'].unique()
# len(df['Reason for Absence'].unique())
# sorted(df['Reason for Absence'].unique())
```

### dattime
```
df_reason_mod['Date'] = pd.to_datetime(df_reason_mod['Date'], format = '%d/%m/%Y')
df_reason_mod['Date'][0].month


def date_to_weekday(date_value):
    return date_value.weekday()

df_reason_mod['Day of the Week'] = df_reason_mod['Date'].apply(date_to_weekday)

```

**Comment by kurtzace** [comment]
## grouping

Group 1: Columns 1 to 14➢Group 2: Columns 15, 16, and 17➢Group 3: Columns 18, 19, 20, and 21➢Group 4: Columns 22 to 28•

```
r1=reasons.loc[:,1:14].max()
r2=reasons.loc[:,15:17].max()
r3=reasons.loc[:,18:21].max()
r4=reasons.loc[:,22:28].max()
```

## to find excess of something (more than average)
```
# find the median of 'Absenteeism Time in Hours'
data_preprocessed['Absenteeism Time in Hours'].median()

data_preprocessed['Excessive Absenteeism'] = np.where(data_preprocessed['Absenteeism Time in Hours'] > 
                   data_preprocessed['Absenteeism Time in Hours'].median(), 1, 0)
```

## to balance above dataset
# check if dataset is balanced (what % of targets are 1s)
# targets.sum() will give us the number of 1s that there are
# the shape[0] will give us the length of the targets array
data_preprocessed['Excessive Absenteeism'] .sum() / data_preprocessed['Excessive Absenteeism'] .shape[0]
